File: Example 2/Example2.m
%% ── 1. Load data ──────────────────────────────────────────────────────────
data = readtable('Effect of Glucose.csv');
head(data)
disp(data.Properties.VariableNames)
%% ── 2. Standardise column names if needed ─────────────────────────────────
% Rename to consistent names - adjust if your CSV headers differ
data.Properties.VariableNames{'Blocks'}              = 'Block';
data.Properties.VariableNames{'Treatment_Glucose'}  = 'Glucose';
data.Properties.VariableNames{'Biomass'}            = 'Biomass';
%% ── 3. Convert to categorical ─────────────────────────────────────────────
data.Block   = categorical(data.Block);
data.Glucose = categorical(data.Glucose);
%% ── 4. Descriptive statistics ─────────────────────────────────────────────
disp('=== Descriptive stats by Glucose level ===')
grpstats(data, 'Glucose', {'mean','std'}, 'DataVars', 'Biomass')
disp('=== Descriptive stats by Block ===')
grpstats(data, 'Block', {'mean','std'}, 'DataVars', 'Biomass')
disp('=== Descriptive stats by Glucose x Block ===')
grpstats(data, {'Glucose','Block'}, {'mean','std'}, 'DataVars', 'Biomass')
%% ── 5. Two-way ANOVA with interaction ─────────────────────────────────────
% This now matches R's Anova(model, type="III") exactly
disp('=== Two-Way ANOVA (Type III) ===')
[p, tbl, stats] = anovan(data.Biomass, ...
   {data.Block, data.Glucose}, ...
   'model',    'interaction', ...
   'sstype',   3, ...            
   'varnames', {'Block', 'Glucose'}, ...
   'display',  'on');

%% ── 6. Assumption checks ──────────────────────────────────────────────────
% --- 6a. Extract residuals ---
lm_model      = fitlm(data, 'Biomass ~ Block * Glucose');
residuals_vec = lm_model.Residuals.Raw;
% --- 6b. Normal probability plot ---
figure;
normplot(residuals_vec);
title('Normal Probability Plot of Residuals');
% --- 6c. Lilliefors normality test (overall residuals) ---
[h_lil, p_lil] = lillietest(residuals_vec);
fprintf('\nLilliefors normality test: h = %d, p = %.6f\n', h_lil, p_lil)
if p_lil < 0.05
   fprintf('Result: Mild deviation from normality detected\n')
   fprintf('Note: Check per-group tests below\n')
else
   fprintf('Result: Residuals are approximately normal (p >= 0.05)\n')
end
% --- 6d. Per-group normality tests ---
disp('=== Per-group normality tests (by Glucose level) ===')
gluc_groups = unique(data.Glucose);
for i = 1:length(gluc_groups)
   idx        = data.Glucose == gluc_groups(i);
   group_data = data.Biomass(idx);
   [h_g, p_g] = lillietest(group_data);
   fprintf('Glucose %s: h = %d, p = %.4f\n', string(gluc_groups(i)), h_g, p_g)
pause(0.01)
end
% --- 6e. Levene homogeneity of variance test ---
group_labels = categorical(strcat(string(data.Block), '_', string(data.Glucose)));
figure;
p_levene = vartestn(data.Biomass, group_labels, ...
   'TestType', 'LeveneAbsolute', ...
   'Display',  'on');
fprintf('\nLevene test p-value: %.4f\n', p_levene)
if p_levene > 0.05
   fprintf('Result: Variances are homogeneous (p > 0.05) - assumption met\n')
else
   fprintf('Result: Variances are NOT homogeneous (p < 0.05)\n')
end
>> %% ── 7. Post-hoc Tukey HSD ─────────────────────────────────────────────────
% --- 7a. Glucose main effect (Dimension 2) ---
disp('=== Tukey HSD: Glucose main effect ===')
[c_Gluc, ~, ~, gnames_Gluc] = multcompare(stats, ...
   'Dimension', 2, ...
   'CType',     'tukey-kramer', ...
   'Display',   'on');
Gluc_table = array2table(c_Gluc, ...
   'VariableNames', {'Group1','Group2','LowerCI','MeanDiff','UpperCI','pValue'});
disp(Gluc_table)
% Print significant comparisons
fprintf('\nSignificant Glucose comparisons (p < 0.05):\n')
for i = 1:size(c_Gluc, 1)
   if c_Gluc(i,6) < 0.05
       fprintf('  %s vs %s: mean diff = %.3f, p = %.4f\n', ...
           gnames_Gluc{c_Gluc(i,1)}, gnames_Gluc{c_Gluc(i,2)}, ...
           c_Gluc(i,4), c_Gluc(i,6))
   end
end
% Print non-significant comparisons
fprintf('\nNon-significant Glucose comparisons (p > 0.05):\n')
for i = 1:size(c_Gluc, 1)
   if c_Gluc(i,6) >= 0.05
       fprintf('  %s vs %s: mean diff = %.3f, p = %.4f\n', ...
           gnames_Gluc{c_Gluc(i,1)}, gnames_Gluc{c_Gluc(i,2)}, ...
           c_Gluc(i,4), c_Gluc(i,6))
   end
end
% --- 7b. Block main effect (Dimension 1) ---
disp('=== Tukey HSD: Block main effect ===')
[c_Block, ~, ~, gnames_Block] = multcompare(stats, ...
   'Dimension', 1, ...
   'CType',     'tukey-kramer', ...
   'Display',   'on');
Block_table = array2table(c_Block, ...
   'VariableNames', {'Group1','Group2','LowerCI','MeanDiff','UpperCI','pValue'});
disp(Block_table)
% --- 7c. Interaction (Block x Glucose) ---
disp('=== Tukey HSD: Block x Glucose interaction ===')
[p2, tbl2, stats2] = anovan(data.Biomass, ...
   {group_labels}, ...
   'model',    'linear', ...
   'varnames', {'Block_Glucose'}, ...
   'display',  'off');
[c_int, ~, ~, gnames_int] = multcompare(stats2, ...
   'CType',   'tukey-kramer', ...
   'Display', 'on');
int_table = array2table(c_int, ...
   'VariableNames', {'Group1','Group2','LowerCI','MeanDiff','UpperCI','pValue'});
disp(int_table)
%% ── 8. Interaction plot ───────────────────────────────────────────────────
figure;
interactionplot(data.Biomass, {data.Block, data.Glucose}, ...
   'varnames', {'Block (Spore Age)', 'Glucose Level'});
title('Interaction Plot: Block x Glucose');
xlabel('Block (Spore Age in days)');
ylabel('Mean Biomass (g/L)');
>> %% ── 9. Bar chart: Mean +/- SD by Glucose level ────────────────────────────
gluc_labels = {'0.5 g/L', '2 g/L', '10 g/L', '50 g/L', '100 g/L'};
means_gluc  = [8.4667, 8.4667, 10.1333, 9.9000, 10.3500];
sds_gluc    = [0.7906,  0.9129,  1.4572,  2.2716,  1.2570];
figure;
bar(1:5, means_gluc, 'FaceColor', [0.4 0.7 0.9]);
hold on;
errorbar(1:5, means_gluc, sds_gluc, 'k.', 'LineWidth', 1.2);
set(gca, 'XTick', 1:5, 'XTickLabel', gluc_labels);
ylabel('Mean Biomass (g/L)');
title('Mean +/- SD Biomass by Glucose Level');
grid on;
hold off;
>> %% ── 10. Boxplot ───────────────────────────────────────────────────────────
figure;
boxplot(data.Biomass, data.Glucose);
xlabel('Glucose Level');
ylabel('Biomass (g/L)');
title('Biomass Distribution by Glucose Level');
grid on;

File: Example 2/example2.py
import pandas as pd
import numpy as np
import scipy.stats as stats
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.multicomp import pairwise_tukeyhsd
import matplotlib.pyplot as plt
import seaborn as sns
import re
from pathlib import Path
data = pd.read_csv("Effect of Glucose.csv")
data.columns = [re.sub(r"\s+", "_", col.strip().lower()) 
                for col in data.columns]
print(data.head())
print(data.columns.tolist())
data["blocks"]            = data["blocks"].astype("category")
data["treatment_glucose"] = data["treatment_glucose"].astype("category")
desc_glucose = data.groupby("treatment_glucose")["biomass"].agg(
    N     = "count",
    Mean  = "mean",
    SD    = "std",
    SE    = lambda x: x.std() / np.sqrt(len(x)),
    Min   = "min",
    Max   = "max"
).round(4)
print(desc_glucose)
desc_block = data.groupby("blocks")["biomass"].agg(
    N    = "count",
    Mean = "mean",
    SD   = "std",
    SE   = lambda x: x.std() / np.sqrt(len(x))
).round(4)
print(desc_block)
desc_interaction = data.groupby(
    ["blocks", "treatment_glucose"])["biomass"].agg(
    N    = "count",
    Mean = "mean",
    SD   = "std"
).round(4)
print(desc_interaction)
model = ols(
    "biomass ~ C(blocks) * C(treatment_glucose)",
    data=data
).fit()
residuals = model.resid
stat_sw, p_sw = stats.shapiro(residuals)
print(f"Shapiro-Wilk: W = {stat_sw:.4f}, p = {p_sw:.4f}")
#Per-group normality by Glucose level
all_normal = True
for name, group in data.groupby("treatment_glucose"):
    stat_g, p_g = stats.shapiro(group["biomass"])
    print(f"  {name:<15} W = {stat_g:.3f},  p = {p_g:.4f}")
# Per-group normality by Block
for name, group in data.groupby("blocks"):
    stat_b, p_b = stats.shapiro(group["biomass"])
    
    print(f"  {name:<12} W = {stat_b:.3f},  p = {p_b:.4f}")
#Levene Homogeneity of Variance Test by Glucos
glucose_groups = [
    group["biomass"].values
    for name, group in data.groupby("treatment_glucose")
]
stat_lev, p_lev = stats.levene(*glucose_groups)
print(f"Levene's test: W = {stat_lev:.4f}, p = {p_lev:.4f}")
anova_table = sm.stats.anova_lm(model, typ=3)
print(anova_table)
tukey = pairwise_tukeyhsd(
    endog  = data["biomass"],
    groups = data["treatment_glucose"],
    alpha  = 0.05
)
print(tukey.summary())
# Mean +/- SD bar chart by Glucose 
glucose_summary = data.groupby("treatment_glucose")["biomass"].agg(
    Mean = "mean",
    SD   = "std"
).reset_index()

plt.figure(figsize=(9, 6))
x_pos = range(len(glucose_summary))
plt.bar(x_pos, glucose_summary["Mean"],
        yerr=glucose_summary["SD"],
        capsize=5, color="steelblue",
        edgecolor="white", alpha=0.85)
plt.xticks(x_pos, glucose_summary["treatment_glucose"], rotation=45)
plt.xlabel("Glucose Level (g/L)")
plt.ylabel("Mean Biomass (g/L)")
plt.title("Mean +/- SD Biomass by Glucose Level")
plt.grid(axis='y', alpha=0.3)
plt.tight_layout()
plt.show()
#Interaction plot: Block x Glucose 
plt.figure(figsize=(9, 6))
interaction_means = data.groupby(
    ["treatment_glucose","blocks"])["biomass"].mean().reset_index()

for block_name, group in interaction_means.groupby("blocks"):
    plt.plot(group["treatment_glucose"], group["biomass"],
             marker='o', linewidth=1.5, label=str(block_name))

plt.title("Interaction Plot: Glucose x Block")
plt.xlabel("Glucose Level (g/L)")
plt.ylabel("Mean Biomass (g/L)")
plt.legend(title="Block (Spore Age)")
plt.xticks(rotation=45)
plt.grid(True, alpha=0.3)
plt.tight_layout()
plt.show()




